### ISSUE TITLE
Accessibility issues in the VidCoder UI

### ISSUE THREAD

**Issue by schulle4u** [RandomEngy/VidCoder#284]
Hello, 

I'm trying to use VidCoder with a screen reader and stumbled upon a few accessibility issues. Many buttons in the main UI are not labeled or display confusing text strings. For example the quality preset selector shows the text "VidCoder.ViewModel.PresetViewModel" for each preset instead of the actual preset name, and when converting a DVD the title selector shows "Handbrake.ApplicationServices.Interop.Json.Scan.SourceTitle", which is not that useful for me. In case of unlabeled buttons I can read some elements by using an OCR function provided by my screen reader, but this is only a temporary solution at best. Other controls such as the menu bar are no problem to access. 

System: Win10 Home 64 Bit, VidCoder V2.62 (X64), also tested with latest available beta. 

Blind people probably are not the main audience for video software, but are there any chances to get these accessibility issues fixed? If you want to test this yourself, you can use the free and open-source [NVDA screen reader](https://www.nvaccess.org). 

Thanks in advance
Steffen

**Comment by RandomEngy** [comment]
Thank for the report. I was thinking about this earlier and I thought I
would prioritize other features because I didn't think blind people would
really be interested in converting video from one format to another. Are
you blind? What do you use it for?

Also would there be any reason why the built-in Windows Narrator software
would not be adequate to test?

On Dec 9, 2017 2:14 AM, "Steffen Schultz" <notifications@github.com> wrote:

Hello,

I'm trying to use VidCoder with a screen reader and stumbled upon a few
accessibility issues. Many buttons in the main UI are not labeled or
display confusing text strings. For example the quality preset selector
shows the text "VidCoder.ViewModel.PresetViewModel" for each preset instead
of the actual preset name, and when converting a DVD the title selector
shows "Handbrake.ApplicationServices.Interop.Json.Scan.SourceTitle", which
is not that useful for me. In case of unlabeled buttons I can read some
elements by using an OCR function provided by my screen reader, but this is
only a temporary solution at best. Other controls such as the menu bar are
no problem to access.

System: Win10 Home 64 Bit, VidCoder V2.62 (X64), also tested with latest
available beta.

Blind people probably are not the main audience for video software, but are
there any chances to get these accessibility issues fixed? If you want to
test this yourself, you can use the free and open-source NVDA screen reader
<https://www.nvaccess.org>.

Thanks in advance
Steffen

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/RandomEngy/VidCoder/issues/284>, or mute the thread

**Comment by schulle4u** [comment]
Yes, I am blind and primarily would like to use VidCoder to create mp4 copies of my DVD collection and have these files centrally accessible on a local network storage device. Especially the possibility to extract multiple audio tracks is great, for example when a DVD contains an additional audio description track for blind people. I've already tested a few tools, but either they have completely inaccessible UI's or their feature set is very limited. Even paid software often ignores blind users when it comes to video-related things. Making VidCoder more accessible therefore not just could make it a good choice for our community, it probably would be the only accessible FOSS solution for us. There is another tool called TEncoder which is accessible in most parts, but I found it to be somewhat unstable and more complicated than VidCoder. 

Windows Narrator has been greatly improved in Windows 10 and of course also can be used to test for accessibility, I just completely forgot about it. 

Steffen

**Comment by RandomEngy** [comment]
Okay, thank you for your patience and insight into the issue! I'm going to put this item on the backlog. I've been doing accessibility work with screen readers and high contrast mode at work, so that should help me out.

I have some larger UI rework planned in the near future, so I will delay accessibility work until that's done so I don't do throwaway work. I would like to get it in before the next stable release though.

**Comment by RandomEngy** [comment]
Okay I've gone over the app top to bottom and fixed a whole raft of issues with screen readers and keyboard navigation in 4.15 Beta.

- All buttons, combo box items, text boxes and number boxes are now all labeled properly
- Save/Rename bar is no longer a focus trap
- You can toggle audio and subtitle tracks with space bar and hear the new toggle value
- You can set focus on the preset list now in the encoding settings window

Unfortunately the formerly accessible menu bar has been replaced by a ribbon UI, which has pretty terrible screen reader support. But I've opened an issue on the Fluent Ribbon project and they are taking a look at it: https://github.com/fluentribbon/Fluent.Ribbon/issues/640 . I hope to have it fixed when they get it resolved on their end.

Please let me know if you find any more usability issues with screen readers.

**Comment by RandomEngy** [comment]
The mystery button you speak of is probably the "Add to Queue" button. Hopefully will be fixed when fluent ribbon updates, along with the Preview, Encode and Stop buttons.

I have a fix ready for the queue items and finished items, so it will read out the source / destination path. That will be in the next Beta version.

The "mystery items" that show up in the scan are probably more Ribbon items. The "group" labels are in the tab order and also there's a unused hidden "tab" on the ribbon that I'm hiding off screen until the update comes in that allows me to hide it properly.

**Comment by RandomEngy** [comment]
Okay the queue items and finished items should be fixed in 4.16 Beta.

**Comment by RandomEngy** [comment]
I've opened a pull request on the Fluent Ribbon project to fix the automation issues with buttons:

https://github.com/fluentribbon/Fluent.Ribbon/pull/646

**Comment by batzen** [comment]
Tracking issues for this on https://github.com/fluentribbon/Fluent.Ribbon/projects/4
If you find more screen reader issues within the ribbon please report them at Fluent.Ribbon.

**Comment by RandomEngy** [comment]
The Ribbon accessibility issues should be fixed in 4.26 Beta, thanks to @batzen 's fixes in the Fluent Ribbon library.

I think we should be 100% good to go now; just let me know if you see any more issues.

**Comment by schulle4u** [comment]
Thank you, looks great so far. :)
The only minor issue I see for the moment is that the ribbon, distinct from normal ribbons in Windows, is not being recocnized as ribbon or menu by the screen reader, and also cannot be opened using the Alt key. A user has to tab around the whole application to access the menu. Some shortcuts would help here, but as said this is only a minor problem. 
The actual problem seems to be completely fixed - thanks again!

Steffen

**Comment by batzen** [comment]
@schulle4u The issues you are talking about are things i started working on, but as i don't know anyone in person who has disabilities i don't know if the things i am implementing will actually help or cause more harm. Would you mind contacting me on gitter so we can chat privately? I would really like to provide early samples to you to get feedback directly. We could also use german in the private chat on gitter ;-)
So if you are interested in improving the screen reader support on the ribbon feel free to contact me any time on gitter.
